lib/runner/pre.js

```diff
@@ -0,0 +1,189 @@
+
+        var triggers = this.triggers,
+            cursor = payload.coords,
+            sourceReadable,
+            resolver,
+            request,
+            mode,
+            data;
+
+        resolver = _.get(this.options, 'fileResolver') ||
+            // todo: remove this once apps give the options correctly
+            _.get(this.options, 'requester.fileResolver');
+
+        // this has changed a bit from the old implementation, a bug is fixed.
+        sourceReadable = resolver && _.isFunction(resolver.createReadStream);
+
+        request = _.get(payload, 'item.request');
+
+        if (!request) { return done(new Error('No request to send.')); }
+
+        // todo: add helper functions in the sdk to do this cleanly for us
+        mode = _.get(request, 'body.mode');
+        data = _.get(request, ['body', mode]);
+
+        // if there is no mode specified, or no data for the specified mode we cannot resolve anything!
+        // @note that if source is not readable, there is no point reading anything, yet we need to warn that file
+        // upload was not done. hence we will have to proceed even without an unreadable source
+        if (!data) { // we do not need to check `mode` here since false mode returns no `data`
+            return done();
+        }
+
+        // in this block, we simply use async.waterfall to ensure that all form of file reading is async. essentially,
+        // we first determine the data mode and based on it pass the waterfall functions.
+        async.waterfall([async.constant(data), {
+            // form data parsing simply "enriches" all form parameters having file data type by replacing / setting the
+            // value as a read stream
+            formdata: function (formdata, next) {
+                // ensure that we only process the file type
+                async.eachSeries(_.filter(formdata.all(), {type: 'file'}), function (formparam, next) {
+                    var stream;
+
+                    try {
+                        sourceReadable && formparam.src && (stream = resolver.createReadStream(formparam.src));
+                    }
+                        // eslint-disable-next-line no-empty
+                    catch (e) { } // we do not need to forward error here since next line we throw warning
+
+                    if (!stream) { // lack of stream also determines lack of `formparam.src` and source being readable
+                        triggers.console(cursor, 'warn', 'unable to load form file for upload: "' + formparam.src + '"');
+                        // @todo - do enabled: false instead of removing it
+                        formdata.remove(formparam); // remove from list since it will interfere with requester
+                    }
+                    // otherwise, replace the value with stream for making it usable in requester
+                    else {
+                        formparam.value = stream;
+                    }
+                    next();
```

Comment:
Potential source of confusion: This callback should be renamed to something other than next.